tokio-reactor/src/lib.rs

```diff
@@ -160,54 +160,41 @@ fn _assert_kinds() {
 
 // ===== impl Reactor =====
 
-/// Set the default reactor for the duration of the closure
-///
-/// # Panics
-///
-/// This function panics if there already is a default reactor set.
-pub fn with_default<F, R>(handle: &Handle, f: F) -> R
-where
-    F: FnOnce() -> R,
-{
-    // Ensure that the executor is removed from the thread-local context
-    // when leaving the scope. This handles cases that involve panicking.
-    struct Reset;
-
-    impl Drop for Reset {
-        fn drop(&mut self) {
-            CURRENT_REACTOR.with(|current| {
-                let mut current = current.borrow_mut();
-                *current = None;
-            });
-        }
-    }
-
-    // This ensures the value for the current reactor gets reset even if there
-    // is a panic.
-    let _r = Reset;
+#[derive(Debug)]
+///Guard that resets current reactor on drop.
+pub struct CurrentReactorReset {}
 
-    CURRENT_REACTOR.with(|current| {
-        {
+impl Drop for CurrentReactorReset {
+    fn drop(&mut self) {
+        CURRENT_REACTOR.with(|current| {
             let mut current = current.borrow_mut();
+            *current = None;
+        });
+    }
+}
 
-            assert!(
-                current.is_none(),
-                "default Tokio reactor already set \
-                 for execution context"
-            );
-
-            let handle = match handle.as_priv() {
-                Some(handle) => handle,
-                None => {
-                    panic!("`handle` does not reference a reactor");
-                }
-            };
+///Sets handle to current reactor, returning guard that unsets it on drop.
+pub fn set_current(handle: &Handle) -> CurrentReactorReset {
```

Comment:
The return value should include the lifetime of `&Handle`. This is to prevent handle from dropping while it is still set as the default.

Reply:
The fn name should also be `set_default` to match the others (or all the others should be changed to `set_current`).

Reply:
Will align all names as `default*`

Reply:
About lifetimes I can do it, but I wonder if it is necessary because we clone them